### ISSUE TITLE
kmod packages for EL8.5

### ISSUE THREAD

**Issue by vstax** [openzfs/zfs#12747]
EL 8.5 was released by Redhat, expected soon to follow by other variants (unfortunately, not Centos, which is what is referred in current repo).

Updated kmods for 8.5 kernel would be appreciated!

**Comment by cyqsimon** [comment]
On a somewhat related note, dkms build on 8.5 is failing. I'm downgrading back to 8.4 for the moment

**Comment by Evernow** [comment]
> unfortunately, not Centos

Not really relevant, but CentOS is getting a 8.5 release.

**Comment by behlendorf** [comment]
For the last release we switched to building packages with almalinux which worked quite well.  When 8.5 is available we'll be able to provide packages.

**Comment by rincebrain** [comment]
> On a somewhat related note, dkms build on 8.5 is failing. I'm downgrading back to 8.4 for the moment

On what version?

When I tried with 4.18.0-348.el8.x86_64 and {2.0.6, 2.1.1, git master (c9d62d13)), with kmod or dkms, following the Building ZFS doc, it built and loaded modules fine...

**Comment by cyqsimon** [comment]
Before the upgrade I had `kernel 4.18.0-305.25.1.el8_4` and `zfs 0.8.6` (for some reason).

I first upgrade the kernel to `4.18.0-348.el8`, for which the dkms build failed. I then realised I was using the incorrect RPM repo for ZFS, so I uninstalled `zfs`, installed the newer repo (`zfs-release.el8_4.noarch.rpm`), and tried installing zfs again, this time `zfs 2.0.6`. But this was still failing.

> On what {kernel, OpenZFS} version?

So tl;dr, {`4.18.0-348.el8`, `2.0.6`}

**Comment by cyqsimon** [comment]
Woah. For some reason a dkms rebuild with `dkms autoinstall -k 4.18.0-348.el8` just compiled with no error this time. I frankly have no idea what happened or what I did that made a difference. I'll edit my OP.

**Comment by mskarbek** [comment]
> For the last release we switched to building packages with almalinux which worked quite well. When 8.5 is available we'll be able to provide packages.

Can't we use RHEL for that? First with current Developer subscription this would be ok, second AFAIK Red Hat provides now special subscriptions for open-source projects. This way, OpenZFS could keep up with RHEL releases.

**Comment by behlendorf** [comment]
> Can't we use RHEL for that?

Redhat has made this easier, but using RHEL is still a bit of a hassle.  The good news is as of this morning [AlmaLinux 8.5](https://almalinux.org/blog/almalinux-os-85-stable-now-available/) was released.

**Comment by mskarbek** [comment]
@behlendorf you don't have to use full RHEL VM for building packages. I'm using `ubi8/ubi` container images. You can register them with `subscription-manager` just like VM and grab all the necessary build deps, build kmod packages, deregister container and discard it.

**Comment by ecnepsnai** [comment]
FYI Rocky 8.5 was released today

**Comment by makhomed** [comment]
> The good news is as of this morning [AlmaLinux 8.5](https://almalinux.org/blog/almalinux-os-85-stable-now-available/) was released.

10 days later:

But kmod packages for EL8.5 still not released - is any estimate exists, when this happens? This is critical issue for us.

**Comment by ecnepsnai** [comment]
@makhomed 

> This is critical issue for us.

You can downgrade the EL release back to 8.4 then wait for the KMOD packages to be released. This has and will happen each time there's a EL release.

**Comment by makhomed** [comment]
@behlendorf
@tonyhutter 

Are there ways to speed up the official build process of ZFS kmod rpms for EL 8.5 and all future versions of Enterprise Linux?

Is there anything I can do to help you with this?

**Comment by mskarbek** [comment]
https://github.com/zeet-cc/zfs-rpms/actions/runs/1499499827 - GH Actions build based on AlmaLinux container. Quick and dirty but can be improved.

**Comment by dhull** [comment]
For anyone who has already upgraded a machine to CentOS 8.5 and needs to get zfs working again, I did so by booting an older CentOS 8.4 kernel. It isn't necessary to downgrade the whole OS to the previous release.

```
grubby --set-default /boot/vmlinuz-4.18.0-305.25.1.el8_4.x86_64
```

**Comment by mskarbek** [comment]
https://github.com/zeet-cc/zfs-rpms/releases/tag/v2.1.1-rhel8.5-4.18.0-348.el8 ;)

**Comment by tonyhutter** [comment]
Hi all, I left for vacation at nearly the same time that Centos 8.5 came out, so sorry for the delay in responding.  I have the EL8.5 kmods and dkms packages built, and am currently testing them.  Hope to have them out soon.

**Comment by tonyhutter** [comment]
The Centos 8.5 kmod RPMs are live now.  Make sure you use the new zfs-release RPM (https://zfsonlinux.org/epel/zfs-release.el8_5.noarch.rpm).  

Thanks all for your patience!

**Comment by stuartthebruce** [comment]
@tonyhutter many thanks. Can you also update the Testing repos with 2.1.1 RPM?

**Comment by tonyhutter** [comment]
@stuartthebruce thanks for reminding me!  Building them now.

**Comment by tonyhutter** [comment]
@stuartthebruce the Centos 8.5 testing repos are live now.

**Comment by stuartthebruce** [comment]
Hmm...am I doing something wrong on this Rocky Linux 8.5 system?
```
[root@zfs3 yum.repos.d]# yum --enablerepo=zfs-testing-kmod install zfs
Last metadata expiration check: 0:08:20 ago on Thu 02 Dec 2021 04:30:39 PM PST.
Dependencies resolved.
======================================================================================================================
 Package                    Architecture           Version                     Repository                        Size
======================================================================================================================
Installing:
 zfs                        x86_64                 2.1.1-1.el8                 zfs-testing-kmod                 653 k
Installing dependencies:
 kmod-zfs                   x86_64                 2.1.1-1.el8                 zfs-testing-kmod                 1.5 M
 libnvpair3                 x86_64                 2.1.1-1.el8                 zfs-testing-kmod                  37 k
 libuutil3                  x86_64                 2.1.1-1.el8                 zfs-testing-kmod                  33 k
 libzfs5                    x86_64                 2.1.1-1.el8                 zfs-testing-kmod                 229 k
 libzpool5                  x86_64                 2.1.1-1.el8                 zfs-testing-kmod                 1.3 M

Transaction Summary
======================================================================================================================
Install  6 Packages

Total download size: 3.7 M
Installed size: 14 M
Is this ok [y/N]: y
Downloading Packages:
(1/6): libuutil3-2.1.1-1.el8.x86_64.rpm                                               268 kB/s |  33 kB     00:00    
(2/6): libnvpair3-2.1.1-1.el8.x86_64.rpm                                              300 kB/s |  37 kB     00:00    
(3/6): libzfs5-2.1.1-1.el8.x86_64.rpm                                                 1.3 MB/s | 229 kB     00:00    
(4/6): zfs-2.1.1-1.el8.x86_64.rpm                                                     2.7 MB/s | 653 kB     00:00    
(5/6): kmod-zfs-2.1.1-1.el8.x86_64.rpm                                                2.2 MB/s | 1.5 MB     00:00    
(6/6): libzpool5-2.1.1-1.el8.x86_64.rpm                                               1.5 MB/s | 1.3 MB     00:00    
----------------------------------------------------------------------------------------------------------------------
Total                                                                                 3.8 MB/s | 3.7 MB     00:00     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                              1/1 
  Installing       : libnvpair3-2.1.1-1.el8.x86_64                                                                1/6 
  Installing       : libuutil3-2.1.1-1.el8.x86_64                                                                 2/6 
  Installing       : libzfs5-2.1.1-1.el8.x86_64                                                                   3/6 
  Installing       : libzpool5-2.1.1-1.el8.x86_64                                                                 4/6 
  Installing       : zfs-2.1.1-1.el8.x86_64                                                                       5/6 
  Running scriptlet: zfs-2.1.1-1.el8.x86_64                                                                       5/6 
  Installing       : kmod-zfs-2.1.1-1.el8.x86_64                                                                  6/6 
  Running scriptlet: kmod-zfs-2.1.1-1.el8.x86_64                                                                  6/6 
  Running scriptlet: zfs-2.1.1-1.el8.x86_64                                                                       6/6 
  Running scriptlet: kmod-zfs-2.1.1-1.el8.x86_64                                                                  6/6 
  Verifying        : kmod-zfs-2.1.1-1.el8.x86_64                                                                  1/6 
  Verifying        : libnvpair3-2.1.1-1.el8.x86_64                                                                2/6 
  Verifying        : libuutil3-2.1.1-1.el8.x86_64                                                                 3/6 
  Verifying        : libzfs5-2.1.1-1.el8.x86_64                                                                   4/6 
  Verifying        : libzpool5-2.1.1-1.el8.x86_64                                                                 5/6 
  Verifying        : zfs-2.1.1-1.el8.x86_64                                                                       6/6 

Installed:
  kmod-zfs-2.1.1-1.el8.x86_64  libnvpair3-2.1.1-1.el8.x86_64 libuutil3-2.1.1-1.el8.x86_64 libzfs5-2.1.1-1.el8.x86_64
  libzpool5-2.1.1-1.el8.x86_64 zfs-2.1.1-1.el8.x86_64       

Complete!
[root@zfs3 yum.repos.d]# zpool list
The ZFS modules are not loaded.
Try running '/sbin/modprobe zfs' as root to load them.
[root@zfs3 yum.repos.d]# modprobe zfs
modprobe: ERROR: ctx=0x5611c0eed2a0 path=/lib/modules/4.18.0-348.2.1.el8_5.x86_64/weak-updates/unicode/zunicode/zunicode.ko.xz error=No such file or directory
modprobe: ERROR: ctx=0x5611c0eed2a0 path=/lib/modules/4.18.0-348.2.1.el8_5.x86_64/weak-updates/unicode/zunicode/zunicode.ko.xz error=No such file or directory
modprobe: ERROR: could not insert 'zfs': Unknown symbol in module, or unknown parameter (see dmesg)
```

**Comment by stuartthebruce** [comment]
I have reproduced this on another Rocky Linux 8.5 test system that had `zfs-dkms-2.1.1-1.el8` installed. After "yum remove zfs" and "yum --enablerepo=zfs-testing-kmod install zfs" I get,
```
[root@zfs8 ~]# modprobe zfs
modprobe: ERROR: ctx=0x5560f4e5f2a0 path=/lib/modules/4.18.0-348.2.1.el8_5.x86_64/weak-updates/unicode/zunicode/zunicode.ko.xz error=No such file or directory
```

**Comment by Evernow** [comment]
> I have reproduced this on another Rocky Linux 8.5 test system that had `zfs-dkms-2.1.1-1.el8` installed. After "yum remove zfs" and "yum --enablerepo=zfs-testing-kmod install zfs" I get,
> 
> ```
> [root@zfs8 ~]# modprobe zfs
> modprobe: ERROR: ctx=0x5560f4e5f2a0 path=/lib/modules/4.18.0-348.2.1.el8_5.x86_64/weak-updates/unicode/zunicode/zunicode.ko.xz error=No such file or directory
> ```

Can you reproduce this on RHEL or Alma?

**Comment by stuartthebruce** [comment]
> Can you reproduce this on RHEL or Alma?

Not easily. Can anyone else confirm that they are able to get `kmod-zfs-2.1.1-1.el8.x86_64` working on an EL8.5 system?

**Comment by rincebrain** [comment]
Rocky 8.5 works fine for me with 4.18.0-348.2.1.el8_5.x86_64 and kmod-zfs-2.1.1-1.el8.x86_64...

**Comment by stuartthebruce** [comment]
`PEBKAC` with the problem became apparent after some cleanup that got me to,
```
[root@zfs3 ~]# yum --enablerepo=zfs-testing-kmod install zfs
Last metadata expiration check: 0:15:18 ago on Sat 04 Dec 2021 09:19:39 AM PST.
Error: 
 Problem: package zfs-2.1.1-1.el8.x86_64 requires zfs-kmod = 2.1.1, but none of the providers can be installed
  - cannot install the best candidate for the job
  - nothing provides kernel(kmalloc_caches) = 0x8200293d needed by kmod-zfs-2.1.1-1.el8.x86_64
  - nothing provides kernel(__kernel_fpu_begin) = 0x0cdb7d12 needed by kmod-zfs-2.1.1-1.el8.x86_64
  - nothing provides kernel(__kernel_fpu_end) = 0x06c0dae5 needed by kmod-zfs-2.1.1-1.el8.x86_64
  - nothing provides kernel(check_disk_change) = 0xf083836a needed by kmod-zfs-2.1.1-1.el8.x86_64
  - nothing provides kernel(revalidate_disk) = 0x6063740c needed by kmod-zfs-2.1.1-1.el8.x86_64
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
```
at which point I realized I was trying to install from the `8.4` repo rather than `8.5`.

**Comment by larsonmattr** [comment]
With Rocky 8.5, I am still getting kmod-zfs 2.0.6-1.el8 with the ifs-release.el8_5.noarch.rpm, and this appears to provide modules for an older kernel 4.18.0-305.19.1, not the active 4.18.0-348.12.2.

Do I need a newer package zfs-2.1.1-1.el8.x86_64, to support this kernels 4.18.0-348.7.1.el8_5 ?  

If so, how or when will it be possible to get this package on Rocky 8.5?

**Comment by cyqsimon** [comment]
@larsonmattr Try uninstalling zfs, removing the existing repo file from `/etc/yum.repos.d/`, then re-adding the repo and installing from a clean state.

See [installation guide](https://openzfs.github.io/openzfs-docs/Getting%20Started/RHEL-based%20distro/index.html#kabi-tracking-kmod)

> When updating to a new RHEL/CentOS minor release the existing kmod packages may not work due to upstream kABI changes in the kernel. After upgrading users must uninstall OpenZFS and then reinstall it from the matching repository as described in this section.

**Comment by filonovd** [comment]
> With Rocky 8.5, I am still getting kmod-zfs 2.0.6-1.el8 with the ifs-release.el8_5.noarch.rpm, and this appears to provide modules for an older kernel 4.18.0-305.19.1, not the active 4.18.0-348.12.2.

You should be getting 2.0.7-1.el8 on your system. It's in the zfs-kmod repo and works fine with 4.18.0-348.12.2.el8_5.x86_64 kernel.
2.1.1-1.el8 is available in the zfs-testing-kmod repo.
